src/js_native_api_v8.h

```diff
@@ -296,49 +287,35 @@ inline v8::Local<v8::Value> V8LocalValueFromJsValue(napi_value v) {
 
 // Adapter for napi_finalize callbacks.
 class Finalizer {
- public:
-  // Some Finalizers are run during shutdown when the napi_env is destroyed,
-  // and some need to keep an explicit reference to the napi_env because they
-  // are run independently.
-  enum EnvReferenceMode { kNoEnvReference, kKeepEnvReference };
-
  protected:
   Finalizer(napi_env env,
             napi_finalize finalize_callback,
             void* finalize_data,
-            void* finalize_hint,
-            EnvReferenceMode refmode = kNoEnvReference)
-      : _env(env),
-        _finalize_callback(finalize_callback),
-        _finalize_data(finalize_data),
-        _finalize_hint(finalize_hint),
-        _has_env_reference(refmode == kKeepEnvReference) {
-    if (_has_env_reference) _env->Ref();
-  }
+            void* finalize_hint)
+      : env_(env),
+        finalize_callback_(finalize_callback),
+        finalize_data_(finalize_data),
+        finalize_hint_(finalize_hint) {}
 
-  ~Finalizer() {
-    if (_has_env_reference) _env->Unref();
-  }
+  virtual ~Finalizer() = default;
 
  public:
   static Finalizer* New(napi_env env,
                         napi_finalize finalize_callback = nullptr,
                         void* finalize_data = nullptr,
-                        void* finalize_hint = nullptr,
-                        EnvReferenceMode refmode = kNoEnvReference) {
-    return new Finalizer(
-        env, finalize_callback, finalize_data, finalize_hint, refmode);
+                        void* finalize_hint = nullptr) {
+    return new Finalizer(env, finalize_callback, finalize_data, finalize_hint);
   }
 
-  static void Delete(Finalizer* finalizer) { delete finalizer; }
+  napi_finalize callback() { return finalize_callback_; }
+  void* data() { return finalize_data_; }
+  void* hint() { return finalize_hint_; }
 
  protected:
```

Comment:
Should we make fields private and use the new accessor methods instead?

Reply:
Not really. These field values are modified in the subclasses. Marking them as private can introduce many unrelated changes. I'll do the field clean-up in a follow-up PR.